sparta/sparta/resources/Queue.hpp

```diff
@@ -67,81 +67,179 @@ namespace sparta
+        uint32_t rollPhysicalIndex_(const uint32_t phys_idx) const {
+            return (vector_size_ - 1) & (phys_idx);
+        }
+
+        /// Utility to convert a given logical index to a physical one
+        uint32_t getPhysicalIndex_(const uint32_t logical_idx) const {
+            return rollPhysicalIndex_(logical_idx + current_head_idx_);
+        }
+
         /// Increment an index value. It is important to make sure
         /// that the index rolls over to zero if it goes above the
         /// bounds of our list.
-        uint32_t incrementIndexValue_(uint32_t val) const
-        {
-            val = (vector_size_-1)&(val+1);
-            return val;
+        uint32_t incrementIndexValue_(const uint32_t val) const {
+            return rollPhysicalIndex_(val + 1);
         }
 
         /// Decrement an index value. It is important to make sure
         /// that the index rolls over to zero if it goes above the
         /// bounds of our list.
-        uint32_t decrementIndexValue_(uint32_t val) const
-        {
-            val = (vector_size_-1)&(val-1);
-            return val;
+        uint32_t decrementIndexValue_(const uint32_t val) const {
+            return rollPhysicalIndex_(val - 1);
         }
 
-        uint32_t offsetIndexBy_(uint32_t val, int offset) const
+        /// The QueueData
+        struct QueueData {
+            template<class DataTRef>
+            QueueData(DataTRef && dat, uint64_t in_obj_id) :
+                data(std::forward<DataTRef>(dat)),
+                obj_id(in_obj_id)
+            {}
+
+            DataT data;
+            uint64_t obj_id = 0;
+        };
+
+        /// Read the data at the physical localation, const
+        const value_type & readPhysical_(const uint32_t phys_idx) const
         {
-            return (vector_size_-1)&(val+offset);
+            return queue_data_[phys_idx].data;
         }
 
-        /// Determine whether or not an index value is currently
-        /// within the bounds of valid objects
-        bool isIndexInValidRange_(uint32_t idx)const
+        /// Access the data at the physical localation, non-const
+        value_type & accessPhysical_(const uint32_t phys_idx)
         {
-            // We convert idx to the user's perception of idx.
-            // then make sure it is smaller than the total_valid_ entries the queue holds
-            uint32_t check = offsetIndexBy_(idx, -current_zero_pos_);
-            if(check < total_valid_){
-                return true;
+            return queue_data_[phys_idx].data;
+        }
+
+        /// Convert the physical index to the logical one
+        uint32_t physicalToLogical_(const uint32_t physical_idx) const {
+            if(physical_idx == invalid_index_) { return invalid_index_; }
+
+            uint32_t logical_idx = 0;
+            uint32_t idx = current_head_idx_;
+            while(idx != physical_idx) {
+                idx = incrementIndexValue_(idx);
+                ++logical_idx;
```

Comment:
Ok, I have a better way to do this based on what Stephen did a million years ago...